schema_utils_test.py

```diff
@@ -474,6 +481,94 @@ def test_get_raises_invalid_validator_id(self):
+            'is_valid_asciimath_expression')
+
+        self.assertTrue(is_valid_asciimath_expression('a+b'))
+        self.assertTrue(is_valid_asciimath_expression('1 + (2*a)'))
+        self.assertTrue(is_valid_asciimath_expression('(a+ b) '))
+        self.assertTrue(is_valid_asciimath_expression('{a+(beta - gamma)}'))
+        self.assertTrue(is_valid_asciimath_expression('(a) / ((b)/(c))'))
+        self.assertTrue(is_valid_asciimath_expression('{a+(b-[c])-(delta^4)}'))
+        self.assertTrue(is_valid_asciimath_expression('alpha + (-3)'))
+        self.assertTrue(is_valid_asciimath_expression('alpha^(3.9/beta*gamma)'))
+        self.assertTrue(is_valid_asciimath_expression('{a-(-3)/(2-(-b)^4)}^2'))
+
+        self.assertFalse(is_valid_asciimath_expression('a+b/'))
+        self.assertFalse(is_valid_asciimath_expression('(352+)-3*x'))
+        self.assertFalse(is_valid_asciimath_expression('42 - [5/a] (4)'))
+        self.assertFalse(is_valid_asciimath_expression('(a-2^34-)'))
+        self.assertFalse(is_valid_asciimath_expression('(25 + 3.4.3*a)'))
+        self.assertFalse(is_valid_asciimath_expression('a_b'))
+        self.assertFalse(is_valid_asciimath_expression('!/'))
+        self.assertFalse(is_valid_asciimath_expression('1+2'))
+        self.assertFalse(is_valid_asciimath_expression('a*b)'))
+        self.assertFalse(is_valid_asciimath_expression('(a}+{b)'))
+        self.assertFalse(is_valid_asciimath_expression('{a+b)(c}'))
+        self.assertFalse(is_valid_asciimath_expression('a**b'))
+        self.assertFalse(is_valid_asciimath_expression('(a)^/(b)'))
+        self.assertFalse(is_valid_asciimath_expression('a+-3'))
+        self.assertFalse(is_valid_asciimath_expression('a=b'))
+        self.assertFalse(is_valid_asciimath_expression('a<b'))
+        self.assertFalse(is_valid_asciimath_expression('a>b'))
+        self.assertFalse(is_valid_asciimath_expression('a<=b'))
+        self.assertFalse(is_valid_asciimath_expression('a>=b'))
+
+    def test_is_valid_numeric_expression_validator(self):
+        """Tests for the is_valid_asciimath_expression static method with
+        numeric type.
+        """
+        is_valid_asciimath_expression = schema_utils.get_validator(
+            'is_valid_asciimath_expression')
+
+        self.assertTrue(is_valid_asciimath_expression('3+2', False))
+        self.assertTrue(is_valid_asciimath_expression('3+2^3', False))
+        self.assertTrue(is_valid_asciimath_expression('(5-2^[6+3])', False))
+        self.assertTrue(is_valid_asciimath_expression('(-5)^(-1)/2', False))
+
+        self.assertFalse(is_valid_asciimath_expression('3+2*a', False))
+        self.assertFalse(is_valid_asciimath_expression('(3+2]', False))
+        self.assertFalse(is_valid_asciimath_expression('3!2', False))
+        self.assertFalse(is_valid_asciimath_expression('3-+2', False))
+        self.assertFalse(is_valid_asciimath_expression('3-5=(-2)', False))
```

Comment:
No examples with ~ ?

Also, in general, these feel like really weak tests that miss a lot of the cases. I don't see greek letters below, I don't see cases that mix latin and greek letters, you'll want to try some equations with spaces in them, etc. Try to be (much!) more comprehensive and think of all the different ways students could write something. I would expect probably three times as many tests as are currently written here, that try to catch a variety of different classes of errors.

Reply:
> No examples with ~ ?

The student isn't expected to enter ~. That would be an invalid character and the expression would be evaluated as an invalid one if it contains ~. The one that has been used in the code is simply to distinguish between the unary negation and binary subtraction, this symbol is arbitrary.

> I don't see greek letters below, I don't see cases that mix latin and greek letters, you'll want to try some equations with spaces in them, etc.

Umm, this is the numeric expression validation which shouldn't contain greek/english alphabets. I have added cases that cover the things you're talking about in the test function right [above](https://github.com/oppia/oppia/pull/9336/files#diff-c74be67925376344e4088b160b1135b4R491). Also, the reason for not adding too many cases here is that this is just a slight variation of the algebraic validation and almost uses the same approach so I thought adding a lot of cases in the algebraic expression test function would indirectly imply that this one works correctly as well. But perhaps that was wrong of me, I'll add more test cases here and in the equation validation as well.

Reply:
Thanks, the tests are better now. And, you're testing interface, not implementation, so all the functions should be treated as independent. The tests for one function don't imply anything about the tests for other functions.

Also, although the student isn't expected to enter ~, they still might, so include some tests with that.

Reply:
Done.